vm/src/sequence.rs

```diff
@@ -1,107 +1,298 @@
+                }
+            }
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn cmp(&'a self, vm: &VirtualMachine, other: &'a Self, op: PyComparisonOp) -> PyResult<bool> {
+        let less = match op {
+            PyComparisonOp::Eq => return self.eq(vm, other),
+            PyComparisonOp::Ne => return self.eq(vm, other).map(|eq| !eq),
+            PyComparisonOp::Lt | PyComparisonOp::Le => true,
+            PyComparisonOp::Gt | PyComparisonOp::Ge => false,
+        };
+
+        let lhs = self.iter();
+        let rhs = other.iter();
+        let lhs_len = lhs.len();
+        let rhs_len = rhs.len();
+        for (a, b) in lhs.zip(rhs) {
+            let ret = if less {
+                vm.bool_seq_lt(a, b)?
+            } else {
+                vm.bool_seq_gt(a, b)?
+            };
+            if let Some(v) = ret {
+                return Ok(v);
+            }
+        }
+        Ok(op.eval_ord(lhs_len.cmp(&rhs_len)))
+    }
 }
 
-impl<'a, D> SimpleSeq for D
-where
-    D: 'a + std::ops::Deref<Target = [PyObjectRef]>,
-{
-    fn len(&self) -> usize {
-        self.deref().len()
+impl<'a> ObjectSequenceOp<'a> for [PyObjectRef] {
+    type Iter = core::slice::Iter<'a, PyObjectRef>;
+
+    fn iter(&'a self) -> Self::Iter {
+        self.iter()
     }
+}
 
-    fn boxed_iter(&self) -> DynPyIter {
-        Box::new(self.deref().iter())
+impl<'a> ObjectSequenceOp<'a> for VecDeque<PyObjectRef> {
+    type Iter = std::collections::vec_deque::Iter<'a, PyObjectRef>;
+
+    fn iter(&'a self) -> Self::Iter {
+        self.iter()
     }
 }
 
-pub(crate) fn eq(vm: &VirtualMachine, zelf: DynPyIter, other: DynPyIter) -> PyResult<bool> {
-    if zelf.len() == other.len() {
-        for (a, b) in Iterator::zip(zelf, other) {
-            if !vm.identical_or_equal(a, b)? {
-                return Ok(false);
+pub trait MutObjectSequenceOp<'a> {
+    type Guard;
+    fn do_get(index: usize, guard: &Self::Guard) -> Option<&PyObjectRef>;
+    fn do_lock(&'a self) -> Self::Guard;
+
+    fn mut_count(&'a self, vm: &VirtualMachine, needle: &PyObject) -> PyResult<usize> {
+        let mut count = 0;
+        self._mut_iter_equal_skeleton::<_, false>(vm, needle, 0..isize::MAX as usize, || {
+            count += 1
+        })?;
+        Ok(count)
+    }
+
+    fn mut_index_range(
+        &'a self,
+        vm: &VirtualMachine,
+        needle: &PyObject,
+        range: Range<usize>,
+    ) -> PyResult<Optioned<usize>> {
+        self._mut_iter_equal_skeleton::<_, true>(vm, needle, range, || {})
+    }
+
+    fn mut_index(&'a self, vm: &VirtualMachine, needle: &PyObject) -> PyResult<Optioned<usize>> {
+        self.mut_index_range(vm, needle, 0..isize::MAX as usize)
+    }
+
+    fn mut_contains(&'a self, vm: &VirtualMachine, needle: &PyObject) -> PyResult<bool> {
+        self.mut_index(vm, needle).map(|x| x.is_some())
+    }
+
+    #[inline]
+    fn _mut_iter_equal_skeleton<F, const SHORT: bool>(
```

Comment:
This method seems somewhat complex despite it declared as `inline`. How do you think?

Reply:
This one is the skeleton for other related function, only few place is calling it. Also as it is generic, so it should be duplicate for each generic parameters. I mark it inline to try to hint the compiler to optimize out the const bool condition check and the closure it called. I don't know what exactly code the compiler produce but inline is what I expected.